server/src/instant/grouped_queue.clj

```diff
@@ -61,15 +92,20 @@
 
        (= t :group-key)
        (let [group-key arg
-             item (first (get @group-key->subqueue group-key))]
+             marked (locking group-key->subqueue
+                      (swap! group-key->subqueue update group-key inflight-queue-pop))
```

Comment:
My thinking is, `process-polling` can take a fn lick `reserve-fn`, which is takes an `group-key` and an `inflight-queue`, and must return a new `inflight-queue`. 

This way, the caller can decide how much to drain the queue. 

We can than pass the reuslt on to `process-fn`